### ISSUE TITLE
Investigate BuildKit options to reduce resources consumption 

### ISSUE THREAD

**Issue by ghassanmas** [overhangio/tutor-mfe#125]
## Context 
By palm it's expected that tutor/tutor-mfe would require BuildKit to be enabled by docker _which is the case by default for docker since  23 version that is BuildKit is the default builder[^1]_.

Buildkit adds extra features to tutor/tutor-mfe, mainly cache related, however one of it's main feature.  
> Parallelize building independent build stages [^1].

Would consume a lot of resouces in case of tutor-mfe, given it would run `npm install` and `npm run build` concurrently for the X MFEs that are enabled by tutor-mfe, this can lead to errors related to network for former and the high resouces consumption for the latter. 

Also another conern about this is that consider the case of which the same machine that is used to deploy an Open edX instance is used for building, it's would be quite risky a low resources machine to run build the image while also having tutor containers running. i.e. in case system crash, it would the affect the availability of the service. 

## Possible solution: 

Note: Those are not exclusive of each others. 

1. Configure BuildKit to use less resources as suggested by @regisb [^2] [^3]
2. Make it optinal to use the BuildKit builder just when building tutor-mfe. 
3. Rethink the way MFEs are built/deployed, i.e.. 
   -  Use External service to build each MFE separatably For example this can be a path openedx/wg-devops/issues/14
 
 ## Related issue/concert: 

Also in Development mode, it has been observed that typically a developer would need to work on a sepsefic MFE, however tutor dev  would by default run all MFEs in development mode, i.e. `npm run start` X times of the enabled MFEs, while is totally different issue, it's probably related.  
 
[^1]: https://docs.docker.com/build/buildkit/#overview 
[^2]: Docs https://docs.docker.com/build/buildkit/configure/#max-parallelism 
[^3]: Slack thread https://openedx.slack.com/archives/CGE253B7V/p1684170597489729

**Comment by regisb** [comment]
Can you please try the following solution?

1. Create a buildtkit.toml configuration file with the following contents:

```
[worker.oci]
   max-parallelism = 2
```

2. Create a builder that uses that configuration:

```
docker buildx create --use --name=max2cpu --driver=docker-container --config=./buildkit.toml
```

3. Build the mfe image:
    
    tutor images build mfe

This solution seems to work for me, in the sense that I see only two layers built simultaneously.

If this works for you as well, then I think that it should be the recommended approach, and we should add these instructions to the troubleshooting docs.

**Comment by ghassanmas** [comment]
Yes @regisb, it did what is expected, I can it only does two thing at a time. 
<img width="1230" alt="image" src="https://github.com/overhangio/tutor-mfe/assets/16361375/a7fad8f1-0c5c-496a-a775-cc2a834f498e">

**Comment by arbrandes** [comment]
I second @regisb's proposal.  A nice and simple way to reduce resource usage.

**Comment by ghassanmas** [comment]
[Update]

I had to change it to use one worker, when testing on Mac M2 mini with 8GB of RAM: 

while building, running `docker stats`

```
CONTAINER ID   NAME                              CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O         PIDS
8c34f988213e   buildx_buildkit_max2cpu           170.35%   868.6MiB / 7.765GiB   10.92%    470MB / 8.78MB    21.8MB / 3.01GB   46
```

CPU would fluctate between 100-200%,  I/O aroud 800GB,  PIDs can reach up to to 60. That when using one worker. 

The crash error I would get otherwise, **npm killed** something. 

To use one worker, I had to update the file pointed above and then running this command:

`docker buildx create --use --node=max2cpu --driver=docker-container --config=./buildkit.toml` the difference is just using `--node` because name exists.

**Comment by ghassanmas** [comment]
Also after the build is done `docker stats` 
```
CONTAINER ID   NAME                              CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O         PIDS
8c34f988213e   buildx_buildkit_max2cpu           0.00%     1.569GiB / 7.765GiB   20.21%    811MB / 13.8MB    1.33GB / 8.45GB   25
```
Why is it still running, it make sense that CPU is 0% because build is done, however it still consumes a lot of RAM. 

I am not sure what magic does buildx/builder do, but I had to stop it `docker stop 8c34f988213e` 

The builder would initially be inactive `docker buildx ls`  and would not appear show up in `docker ps` or `docker stats` as a running container, until a build command is execuated. The probelm again is even after build is done, the builder container would still be running... may be I had to wait for it to stop itself, I couldn't find a relavnt ref in the doc.

**Comment by regisb** [comment]
> I had to change it to use one worker, when testing on Mac M2 mini with 8GB of RAM

Running with just two workers exceeds your 8GB of RAM??? This would mean that building a single MFE requires 4GB of RAM? If this is true then we really need to rename MFE to macrofrontends.

> Why is it still running, it make sense that CPU is 0% because build is done, however it still consumes a lot of RAM.

Buildx is actually a process that runs inside a docker container -- as implied by the `--driver=docker-container` option you used. I suspect it's using some memory because it's doing [garbage collection](https://docs.docker.com/build/cache/garbage-collection/) and other chores in the background. In my experience it's safe to remove the container, as it will automatically be restarted next time you run `docker buildx`.

**Comment by arbrandes** [comment]
> we really need to rename MFE to macrofrontends.

No objections from me. ;P

**Comment by regisb** [comment]
I changed the title of the issue to reflect the decision proposed in my [earlier comment](https://github.com/overhangio/tutor-mfe/issues/125#issuecomment-1549669393).
